keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    // ObservedGeneration is used by the Application Controller to report the last Generation of an Application 
+    // that it has observed.
+    ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"int64,1,opt,name=observedGeneration"`
+    
+    // Installed is used by the Application Controller ot report the installed Components and Dependencies of  
+    // an Application.
+    Installed []string `json:"installed,omitempty" protobuf:"bytes,2,opt,name=installed"`
+    
+    // Ready is used to report the readiness of an Application. If an Application contains an ApplicationHealthCheck, 
+    // this field corresponds directly to the Readiness of the Pod generated from that health check. If the 
+    // Application does not contain an ApplicationHealthCheck, this field is set to Unknown.
+    Ready ConditionStatus json:`"ready,omitempty" protobuf:"bytes,3,opt,name=ready"`
+}
+
+// ApplicationHealthCheck contains a template for a Pod that performs an Application health check. The readiness 
+// of the Pod decides the readiness of the Application.
+type ApplicationHealthCheck struct {
+    
+    // Template is used to specify a Pod whose readiness will be used to determine the readiness of an Application.
+    Template v1.PodTemplateSpec
+}
+// ApplicationSpec is the specification object for an Application. It specifies the Application's components, 
+// dependencies, and an optional health check. Its selector can be used to select application components.
+type ApplicationSpec struct {
+    // Type is the type of the application (e.g. WordPress, MySQL, Cassandra).
+    Type string `json:"type" protobuf:"bytes,1,name=type"`
+    
+    // Components is a map of the applications components to the kinds created by the application (e.g. Pods, Services, 
+    // Deployments, CRDS, etc)
+    Components map [string]metav1.GroupVersionKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
+    
+    // HealthCheck is an optional application level health check.
+    HealthCheck* ApplicationHealthCheck `json:"healthCheck" protobuf:"bytes,5,opt,name=healthCheck"`
+    
+    // Version is an optional version indicator for the application.
+    Version string `json:"version,omitempty" protobuf:"bytes,6,opt,name=version"`
+    
+    // AboutURL is a URL pointing to any additional information about the application.
+    AboutURL string `json:"url,omitempty" protobuf:"bytes,7,opt,name=url"`
```

Comment:
Why is this necessary? What user requirement is this satisfying? Are there any examples?

Reply:
Commenting generally on descriptive metadata (not just this AboutURL field), the user requirements have to do with publishing and consuming off-the-shelf applications.

For example, as a new Kubernetes user, if I run "helm install stable/wordpress", I'd like to be able to navigate to my favorite k8s dashboard and find a basic landing page about the instance WordPress that I just created. The grouping and health mechanisms in this proposal begin to satisfy this requirement, but what I'd also like to see on such a page is high-level descriptive information, links to documentation, where to find the Admin URL, username, and password, suggested next steps, and similar app-specific content.

As for examples, I would point to Helm's Chart.yaml and NOTES.txt. These documents represent structured descriptive metadata and lightweight instructions. I see this proposal as an inter-operable standard for that same class of information, with the added benefit that the information would be available for rendering in a UI.

By the way, I think this user requirement could be called out more directly in the "User Stories" section.

Reply:
Perhaps we can avoid needing consensus on which fields are sufficiently ubiquitous (and their semantics) by modeling all fields not relevant to the application controller with @deustis's suggestion below (ordered list of key-value pairs).

This model may be insufficient if multiple UI providers require different semantics of the same Application instance, so we might add a nesting layer to differentiate between UI providers (a mapping of UI provider to ordered list of key-value pairs).